packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -27,23 +29,162 @@ import PropTypes from 'prop-types'
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: item?.loaded ?? false
+        }))
+    })
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        const findKeys = Object.keys(objToFindBy)
+
+        const rootIsMatch = findKeys.every((key) => isEqual(tree[key], objToFindBy[key]))
+        if (rootIsMatch) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            Object.assign(tree, replacementObj)
+            return tree
+        }
+
+        const findInChildren = (obj) => {
+            if (!Object.prototype.hasOwnProperty.call(obj, childrenKey)) return false
+            const children = obj[childrenKey]
+
+            const someChildIsMatch = children.some((child, idx, arr) => {
+                const childIsMatch = findKeys.every((key) => {
+                    return isEqual(child[key], objToFindBy[key])
+                })
+                if (childIsMatch) {
+                    arr[idx] = replacementObj
+                }
+                return childIsMatch
+            })
+            if (someChildIsMatch) return true
+
+            // We can't use `.some` for the search because it short-circuits
+            // and we still want to check every child.
+            const childrenModified = children.map((child) => {
+                return findInChildren(child)
+            })
+            // If any child was modified, then so was this `obj`
+            return childrenModified.some(Boolean)
+        }
+
+        const treeModified = findInChildren(tree)
+        return treeModified && tree
+    }
+
+    const fetchCategoryNode = async (id, levels = 1) => {
+        const STALE_TIME = 10000 // 10 seconds
+        // return early if there's one that is less than stale time
+        // than the stale time
+        const storageItem = window?.localStorage?.getItem(LOCAL_STORAGE_PREFIX + id + locale)
+        if (Date.now() < storageItem?.fetchTime + STALE_TIME) {
+            return storageItem
+        }
+        let res
+        try {
+            res = await api.shopperProducts.getCategory({
+                parameters: {
+                    id,
+                    levels
+                }
+            })
+        } catch (e) {
+            return e
+        }
+        const newTree = findAndModifyFirst(
+            root,
+            itemsKey,
+            {id},
+            {
+                ...res,
+                loaded: true,
+                [itemsKey]: res?.[itemsKey]?.map((item) => ({...item, loaded: true}))
+            }
+        )
+        setRoot(newTree)
```

Comment:
Using `setRoot` after `await` doesn't play nice with React. I don't know enough to know the proper solution, but I think this is a good use case for the [useCategory hook](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/1c1ae360ff004ce0c443fdfaf0efbcbc094e0aa9/packages/commerce-sdk-react/src/hooks/ShopperProducts/query.ts#L136). Or you could figure out how to make react place nice with an async function. [This might help?](https://devtrium.com/posts/async-functions-useeffect)